src/osd/ECBackend.cc

```diff
@@ -912,6 +918,29 @@ void ECBackend::handle_sub_read(
 	  );
       }
     }
+    // If all reads happened then lets check digest
+    if (j == i->second.end()) {
+      dout(20) << __func__ << ": Checking hash of " << i->first << dendl;
+      ECUtil::HashInfoRef hinfo = get_hash_info(i->first);
+      // This shows that we still need deep scrub because large enough files
+      // are read in sections, so the digest check here won't be done here.
+      if (!hinfo || (total_read == hinfo->get_total_chunk_size() &&
+	  h.digest() != hinfo->get_chunk_hash(shard))) {
+        if (!hinfo) {
+	  get_parent()->clog_error() << __func__ << ": No hinfo for " << i->first << "\n";
+	  dout(5) << __func__ << ": No hinfo for " << i->first << dendl;
+	} else {
+	  get_parent()->clog_error() << __func__ << ": Bad hash for " << i->first << " digest 0x"
+	          << hex << h.digest() << " expected 0x" << hinfo->get_chunk_hash(shard) << dec << "\n";
+	  dout(5) << __func__ << ": Bad hash for " << i->first << " digest 0x"
+	          << hex << h.digest() << " expected 0x" << hinfo->get_chunk_hash(shard) << dec << dendl;
+	}
+	// Do NOT check osd_read_eio_on_bad_digest here.  We need to report
+	// the state of our chunk in case other chunks could substitute.
+	reply->buffers_read.erase(i->first);
```

Comment:
I've thought about this further and since the new code can read another shard to reconstruct the object, it is better to return EIO for a shard in which the hash info isn't present.  Furthermore, if `get_hash_info()` sees "Mismatch of total_chunk_size" it won't return any hash info.  In that case we would have a bad checksum since the it is based on a differently sized shard.